agent/service_manager.go

```diff
@@ -0,0 +1,237 @@
+	if err := s.startConfigWatch(); err != nil {
+		return err
+	}
+	go s.runWatch()
+
+	return nil
+}
+
+func (s *serviceConfigWatch) Stop() {
+	s.cancelFunc()
+}
+
+// runWatch handles any update events from the cache.Notify until the
+// config watch is shut down.
+func (s *serviceConfigWatch) runWatch() {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case event := <-s.updateCh:
+			if err := s.handleUpdate(event, false); err != nil {
+				s.agent.logger.Printf("[ERR] agent.manager: error handling service update: %v", err)
+			}
+		}
+	}
+}
+
+// handleUpdate receives an update event about either the service registration or the
+// global config defaults, updates the local state and re-registers the service with
+// the newly merged config. This function takes the serviceConfigWatch lock to ensure
+// only one update can be happening at a time.
+func (s *serviceConfigWatch) handleUpdate(event cache.UpdateEvent, locked bool) error {
+	// Take the agent state lock if needed. This is done before the local config watch
+	// lock in order to prevent a race between this config watch and others - the config
+	// watch lock is the inner lock and the agent stateLock is the outer lock.
+	if !locked {
+		s.agent.stateLock.Lock()
+		defer s.agent.stateLock.Unlock()
+	}
+	s.Lock()
+	defer s.Unlock()
+
+	if event.Err != nil {
+		return fmt.Errorf("error watching service config: %v", event.Err)
+	}
+
+	switch event.Result.(type) {
+	case *serviceRegistration:
+		s.registration = event.Result.(*serviceRegistration)
+	case *structs.ServiceConfigResponse:
+		resp := event.Result.(*structs.ServiceConfigResponse)
+		s.config = &resp.Definition
+	default:
+		return fmt.Errorf("unknown update event type: %T", event)
+	}
+
+	service := s.mergeServiceConfig()
+	err := s.agent.addServiceInternal(service, s.registration.chkTypes, s.registration.persist, s.registration.token, s.registration.source)
+	if err != nil {
+		return fmt.Errorf("error updating service registration: %v", err)
+	}
+
+	return nil
+}
+
+// startConfigWatch starts a cache.Notify goroutine to run a continuous blocking query
+// on the resolved service config for this service.
+func (s *serviceConfigWatch) startConfigWatch() error {
+	name := s.registration.service.Service
+
+	req := &structs.ServiceConfigRequest{
+		Name:         name,
+		Datacenter:   s.agent.config.Datacenter,
+		QueryOptions: structs.QueryOptions{Token: s.agent.config.ACLAgentToken},
+	}
+	if s.registration.token != "" {
+		req.QueryOptions.Token = s.registration.token
+	}
+	err := s.agent.cache.Notify(s.ctx, cachetype.ResolvedServiceConfigName, req, fmt.Sprintf("service-config:%s", name), s.updateCh)
+
+	return err
+}
+
+// updateRegistration does a synchronous update of the local service registration and
+// returns the result.
+func (s *serviceConfigWatch) updateRegistration(registration *serviceRegistration) error {
+	return s.handleUpdate(cache.UpdateEvent{
+		Result: registration,
+	}, true)
```

Comment:
We are claiming that the agent stateLock is held here but that is not ensured by the method and it's not mentioned in the comment. It would be racy to call this without holding agent.stateLock right?